aws-serverless-java-container-core/src/main/java/com/amazonaws/serverless/proxy/internal/servlet/AwsHttpServletRequest.java

```diff
@@ -0,0 +1,312 @@
+ * with the License. A copy of the License is located at
+ *
+ * http://aws.amazon.com/apache2.0/
+ *
+ * or in the "license" file accompanying this file. This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES
+ * OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions
+ * and limitations under the License.
+ */
+package com.amazonaws.serverless.proxy.internal.servlet;
+
+import com.amazonaws.services.lambda.runtime.Context;
+
+import javax.servlet.AsyncContext;
+import javax.servlet.DispatcherType;
+import javax.servlet.ServletContext;
+import javax.servlet.http.Cookie;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpSession;
+import java.io.UnsupportedEncodingException;
+import java.net.URLEncoder;
+import java.nio.charset.StandardCharsets;
+import java.util.*;
+
+/**
+ * Base HttpServletRequest object. This object exposes some utility methods to work with request values such as headers
+ * and query string parameters. New implementations of <code>HttpServletRequest</code> can extend this class to reuse
+ * the utility methods
+ */
+public abstract class AwsHttpServletRequest implements HttpServletRequest {
+
+    //-------------------------------------------------------------
+    // Constants
+    //-------------------------------------------------------------
+
+    static final String HEADER_KEY_VALUE_SEPARATOR = "=";
+    static final String HEADER_VALUE_SEPARATOR = ";";
+    static final String FORM_DATA_SEPARATOR = "&";
+    static final String DEFAULT_CHARACTER_ENCODING = "UTF-8";
+    static final String HEADER_DATE_FORMAT = "EEE, d MMM yyyy HH:mm:ss z";
+    static final String ENCODING_VALUE_KEY = "charset";
+
+    // We need this to pickup the protocol from the CloudFront header since Lambda doesn't receive this
+    // information from anywhere else
+    static final String CF_PROTOCOL_HEADER_NAME = "CloudFront-Forwarded-Proto";
+
+
+    //-------------------------------------------------------------
+    // Variables - Private
+    //-------------------------------------------------------------
+
+    private Context lamdaContext;
+    private DispatcherType dispatcherType;
+    private Map<String, Object> attributes;
+
+
+    //-------------------------------------------------------------
+    // Constructors
+    //-------------------------------------------------------------
+
+    /**
+     * Protected constructors for implemnenting classes. This should be called first with the context received from
+     * AWS Lambda
+     * @param lambdaContext The Lambda function context. This object is used for utility methods such as log
+     */
+    AwsHttpServletRequest(Context lambdaContext) {
+        lamdaContext = lambdaContext;
+        attributes = new HashMap<>();
+
+        // TODO: We are setting this to request by default
+        dispatcherType = DispatcherType.REQUEST;
```

Comment:
Changed to static return. Cannot see where we'll use the other dispatcher types any time soon: http://docs.oracle.com/javaee/6/api/javax/servlet/DispatcherType.html